src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,214 @@
+          if (descendantContext[0] === '__EM__') return // preventing nodes at level 0 from merged to this (temporary fix)
+          const longContext = descendantContext.length > newContext.length ? descendantContext : newContext
+          const shortContext = newContext.length < descendantContext.length ? newContext : descendantContext
+          // siblings A.B.C and A.B.D ---> A.B.D
+          if (descendantContext.length === newContext.length && newContext.length - commonPath.length === 1) {
+            isMerged = true
+            _.unset(tree, descendantContext)
+            _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+          }
+          /*
+            restricting merge of direct cousins like A.B.C.F and A.B.D.E
+            merging distant relation A.B.C and A.B.D.E.F.G.H ---> A.B.D.E.F.G.H
+          */
+          else if (!(shortContext.length - commonPath.length > 1 || (newContext.length === descendantContext.length && newContext.length - commonPath.length === 2))) {
+            isMerged = true
+            _.unset(tree, shortContext.slice(0, commonPath.length + 1))
+            _.set(tree, longContext, { leaf: true, lastUpdated: timestamp(), path: newContext.length > descendantContext.length ? newPath : descendant.path })
+          }
+        })
+        // adding new thought to the tree
+        if (!isMerged) _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+      }
+    }
+  }
+}
+
+/**
+   * Deletes node from existing tree object using mutation.
+   * @param {object} tree nested object representing tree
+   * @param {string[]} oldPath array of string representing path
+   * @param {string[]} newPath array of string representing path
+   * @param {boolean} [timestampUpdate=true] if false it doesn't update lastUpdated property of all affected leaf nodes (used by nodeMove)
+   */
+const nodeDelete = (tree, oldPath, timestampUpdate = true) => {
+  const oldContext = pathToContext(oldPath)
+  const { node: commonNode, path: commonPath } = findTreeDeepestSubcontext(tree, oldContext)
+  if (commonNode) {
+    if (oldContext.length > commonPath.length) return // if node is not already available in the tree
+    const closestAncestor = findClosestAncestorWithMultipleChildren(tree, commonPath)
+    const nodeToMergeIntoPath = commonPath.length - closestAncestor.path.length <= 2 ? closestAncestor.path : false
+    if (nodeToMergeIntoPath) {
+      // checking all the nodes from the direct parent or grandparent , if available updating all the descendants lastUpdated
+      const pathBeingMerged = commonPath.slice(0, closestAncestor.path.length + 1)
+      _.unset(tree, pathBeingMerged) // deleting the merged path
+      findTreeDescendants(tree, nodeToMergeIntoPath).forEach(descendant => {
+        _.set(tree, pathToContext(descendant.path), { leaf: true, lastUpdated: timestampUpdate ? timestamp() : descendant.lastUpdated, path: descendant.path })
```

Comment:
I updated it before because `A.B.C.D.E` is the changed path and it has now distant relation with these `A.B.C.G.H.K` and `A.B.C.G.H.I.J` . So it's actually merging to these paths. But I too think we should not update the timestamp in case of delete. Should we change it?

Reply:
I've thought about it some more, and let's leave it for now. There is some reason to it, since   `A.B.C.G.H.I.J` and `A.B.C.G.H.K` are the closest ancestors, and if they are not updated then `A.B.C.*` will not be accessible from the top of the list. I can assess more after actual usage. Thanks!